File: lib/domain/schedule/schedule_window.dart
/// Pure-Dart helper that decides whether a given moment is inside a
/// schedule's active window. Consumed by:
/// - Phase 2 editor preview (LIST-09 visual indicator on the entry row)
/// - Phase 4 PAUS-10 (interception is gated on this returning true)
/// - Phase 5 STRK-09 (streak counting only when the entry is "active")
///
/// Pure Dart — no Flutter, no Riverpod imports.
library;

/// True iff [now] falls within the schedule defined by start/end
/// minutes-of-day and a weekday bitmask.
///
/// The three schedule values are nullable as a triple: an entry has a
/// schedule iff all three are non-null. Returning false when ANY of them is
/// null lets callers treat "no schedule" as "dormant for the purpose of this
/// helper" — they can layer their own always-on policy on top.
///
/// **Same-day windows** (`start <= end`) use a half-open interval
/// `[start, end)`. At `start`, the entry is in the window; at `end`, it is
/// out. This matches typical "from 09:00 to 17:00" intent — at 17:00:00 the
/// window is closed.
///
/// **Cross-midnight windows** (`end < start`, e.g., 22:00 → 06:00) cover
/// `[today's start → tomorrow's end]`. The weekday mask refers to the
/// **START day** for both halves: a "weekdays-only 22:00–06:00" schedule is
/// active Mon-night-into-Tue, …, Fri-night-into-Sat. So at 03:00 on a
/// Saturday, the START day was Friday — we look up `yesterday.weekday`.
///
/// **Streak day boundary (04:00)** is OUT of scope for this helper. Phase 5
/// shifts wall-clock day → streak day via `streakDayFor` separately
/// (see `lib/domain/schedule/streak_day.dart`).
bool isInScheduleWindow({
  required DateTime now,
  required int? startMinutes,
  required int? endMinutes,
  required int? weekdayMask,
}) {
  if (startMinutes == null || endMinutes == null || weekdayMask == null) {
    return false;
  }
  final localNow = now.toLocal();
  final nowMinutes = localNow.hour * 60 + localNow.minute;
  final todayWeekdayBit = 1 << (localNow.weekday - 1); // Mon=1..Sun=7

  if (startMinutes <= endMinutes) {
    // Same-day window
    final inWindow = nowMinutes >= startMinutes && nowMinutes < endMinutes;
    return inWindow && (weekdayMask & todayWeekdayBit) != 0;
  } else {
    // Cross-midnight window
    if (nowMinutes >= startMinutes) {
      // Tail of [start..midnight) on TODAY — START-day weekday must match
      return (weekdayMask & todayWeekdayBit) != 0;
    } else if (nowMinutes < endMinutes) {
      // Tail of [midnight..end) on TODAY — START day was YESTERDAY
      final yesterday = localNow.subtract(const Duration(days: 1));
      final yesterdayBit = 1 << (yesterday.weekday - 1);
      return (weekdayMask & yesterdayBit) != 0;
    }
    return false;
  }
}

File: android/app/src/main/kotlin/com/nottodo/not_to_do_list/service/ScheduleWindow.kt
package com.nottodo.not_to_do_list.service

import java.util.Calendar
import java.util.TimeZone

/**
 * Kotlin port of lib/domain/schedule/schedule_window.dart (Phase 2).
 *
 * D-12: byte-for-byte semantic parity with the Dart helper, pinned by
 * 200+ tuples in:
 *   - test/domain/schedule/schedule_window_parity_test.dart (Dart side)
 *   - android/app/src/test/.../ScheduleWindowTest.kt        (Kotlin side)
 *
 * Called by NotToDoAccessibilityService (Plan 04-05) to gate Intent launches
 * on PAUS-10 schedule windows. Outside the window: launch passes through.
 *
 * Same-day windows (start <= end) use a half-open interval [start, end).
 * Cross-midnight windows (end < start) cover [today.start..tomorrow.end];
 * the weekday mask refers to the START day (so a "weekdays-only 22:00–06:00"
 * schedule is active Mon-night-into-Tue, not Mon-night-into-Mon).
 *
 * PLAY-02: this file does pure time arithmetic; no AccessibilityService API
 * surface. Forbidden tokens are not present.
 */
fun isInScheduleWindow(
    nowEpochMs: Long,
    startMinutes: Int?,
    endMinutes: Int?,
    weekdayMask: Int?,
    timeZone: TimeZone = TimeZone.getDefault(),
): Boolean {
    // Null-guard: all three must be non-null (matches Dart line 38-40)
    if (startMinutes == null || endMinutes == null || weekdayMask == null) return false

    // Convert epoch + timezone to local hour/minute/weekday (mirrors Dart's now.toLocal())
    val cal = Calendar.getInstance(timeZone)
    cal.timeInMillis = nowEpochMs
    val nowMin = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)

    // Java Calendar: SUN=1..SAT=7. Dart: MON=1..SUN=7.
    // Conversion: dartWeekday = ((javaDayOfWeek + 5) % 7) + 1
    val dartToday = ((cal.get(Calendar.DAY_OF_WEEK) + 5) % 7) + 1
    val todayBit = 1 shl (dartToday - 1)

    return if (startMinutes <= endMinutes) {
        // Same-day window [start, end): half-open (matches Dart line 45-49)
        val inWindow = nowMin >= startMinutes && nowMin < endMinutes
        inWindow && (weekdayMask and todayBit) != 0
    } else {
        // Cross-midnight window (end < start)
        if (nowMin >= startMinutes) {
            // Head [start..midnight): check TODAY's weekday bit (START day)
            (weekdayMask and todayBit) != 0
        } else if (nowMin < endMinutes) {
            // Tail [midnight..end): START day was YESTERDAY — check yesterday's bit
            val yCal = Calendar.getInstance(timeZone)
            yCal.timeInMillis = nowEpochMs - 86_400_000L
            val dartYesterday = ((yCal.get(Calendar.DAY_OF_WEEK) + 5) % 7) + 1
            val yBit = 1 shl (dartYesterday - 1)
            (weekdayMask and yBit) != 0
        } else {
            // Gap [end..start): outside the window
            false
        }
    }
}
